cli/command/service/progress/progress.go

```diff
@@ -328,30 +324,73 @@ func (u *replicatedProgressUpdater) update(service swarm.Service, tasks []swarm.
 	return running == replicas, nil
 }
 
-type globalProgressUpdater struct {
-	progressOut progress.Output
-
-	initialized bool
-	done        bool
-}
-
-func (u *globalProgressUpdater) update(service swarm.Service, tasks []swarm.Task, activeNodes map[string]swarm.Node, rollback bool) (bool, error) {
-	// If there are multiple tasks with the same node ID, favor the one
+func (u *replicatedProgressUpdater) tasksBySlot(tasks []swarm.Task, activeNodes map[string]struct{}) map[int]swarm.Task {
+	// If there are multiple tasks with the same slot number, favor the one
 	// with the *lowest* desired state. This can happen in restart
 	// scenarios.
-	tasksByNode := make(map[string]swarm.Task)
+	tasksBySlot := make(map[int]swarm.Task)
 	for _, task := range tasks {
-		if numberedStates[task.DesiredState] == 0 {
+		if numberedStates[task.DesiredState] == 0 || numberedStates[task.Status.State] == 0 {
 			continue
 		}
-		if existingTask, ok := tasksByNode[task.NodeID]; ok {
-			if numberedStates[existingTask.DesiredState] <= numberedStates[task.DesiredState] {
+		if existingTask, ok := tasksBySlot[task.Slot]; ok {
+			if numberedStates[existingTask.DesiredState] < numberedStates[task.DesiredState] {
+				continue
+			}
+			// If the desired states match, observed state breaks
+			// ties. This can happen with the "start first" service
+			// update mode.
+			if numberedStates[existingTask.DesiredState] == numberedStates[task.DesiredState] &&
+				numberedStates[existingTask.Status.State] <= numberedStates[task.Status.State] {
 				continue
 			}
 		}
-		tasksByNode[task.NodeID] = task
+		if task.NodeID != "" {
+			if _, nodeActive := activeNodes[task.NodeID]; nodeActive {
+				tasksBySlot[task.Slot] = task
+			}
```

Comment:
If `task.NodeID` is not empty but the node is not active then we skip the task.

Minor nit, but this case seems like it would be more obvious like this:

```go

if _, nodeActive := activeNodes[task.NodeID]; !nodeActive {
    continue
}

tasksBySlot[task.Slot] = task
```

That way it matches the checks above and there is only one line which assigns to `tasksBySlot`